### DISCUSSION TITLE
SVG support (Namespace Element support)

### DISCUSSION THREAD

**Discussion by andrewgryan** [vanjs-org/van#27] [Ideas]
SVG has a slightly different DOM API for element creation than ordinary DOM elements. Referred to as "namespace" elements in the spec (https://developer.mozilla.org/en-US/docs/Web/API/Document/createElementNS)

I noticed while trying to add icons to an app using VanJS that the element in the devtools `<svg width="100" ...><circle cx="50" ...></circle></svg>` while visually identical to the namespace element didn't render on my page. 

```
// SVG namespace element API
const el = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
el.setAttribute("width", "100")
// etc.
```

Would it be possible to add support for SVG in VanJS without sacrificing the amazing developer experience?

Either switch to `createElementNS` for known namespace tags, e.g. `path, p, circle` etc. or allow users to opt-in somehow?

**Comment by andrewgryan** [comment]
Perhaps ship with a `ns_tags` Proxy similar to this line but with the creation API switch:

https://github.com/vanjs-org/van/blob/3f98060971a8ff141179ef0fcd9f1b13b536a6f2/src/van.js#L49

**Comment by Tao-VanJS** [comment]
Yes. SVG support can be considered. I'm not familiar with native API for creating SVG elements. Could you please share a piece of code with native API and I will think through how to extend VanJS to support it.

**Reply by andrewgryan** [reply]
Sure thing, it's very similar to the standard `document.createElement("tagName")` API but with an extra positional argument, which is a special string related to the type of namespace. Here's the w3schools yellow/green circle example using native API.

```js
const svg = () => {
  const el = document.createElementNS("http://www.w3.org/2000/svg", "svg");
  el.setAttribute("width", "100");
  el.setAttribute("heigth", "100");

  // Make circle
  const c = document.createElementNS("http://www.w3.org/2000/svg", "circle");
  c.setAttribute("cx", "50");
  c.setAttribute("cy", "50");
  c.setAttribute("r", "40");
  c.setAttribute("stroke", "green");
  c.setAttribute("stroke-width", "4");
  c.setAttribute("fill", "yellow");
  el.appendChild(c);

  return el;
};
```

**Reply by Tao-VanJS** [reply]
Thanks! Looks like the nsTags approach is reasonable. Another idea is to have:

```js
const {circle, svg} = tags["http://www.w3.org/2000/svg"]
```

Do all namespaces start with "http"? Need a way to differentiate tag name and namespace.

**Reply by andrewgryan** [reply]
To be honest, I didn't realise SVG tags were a slightly different flavour of DOM element myself until about an hour of trying to get it to work and then having a moment of inspiration to check if it was the `createElement` call causing the problem.

 I'm not sure surfacing the namespace URI into app code is the way to go. Also, I don't think there are too many namespaces kicking about either, so supporting the popular ones with a simple API and an escape hatch for edge cases might work better.

```js
const {circle, svg} = svgTags
```

perhaps?

**Reply by Tao-VanJS** [reply]
The problem with that is a new function needs to be introduced whenever we want to support a new namespace.

**Reply by andrewgryan** [reply]
I guess my idea was to have `svgTags` as a convenient Proxy function for those wishing to add icons/graphics to their apps, e.g. `const svgTags = nsTags("http://www.w3.org/2000/svg")`. The `nsTags` utility can provide a seam for future namespace elements, my feeling is there aren't that many, svg and mathml are examples given in the MDN docs.

`nsTags` can be a closure over the `uri` string that returns a Proxy with the `createElementNS(uri, ...)` swapped in. Hopefully, with some clever refactoring it won't affect the bundle size much.

**Reply by Tao-VanJS** [reply]
Perhaps we should call the function `tagsNS`, to align with the naming convention in DOM.

**Reply by andrewgryan** [reply]
I'd agree with the alignment except for the awkwardness of `tagsSVG`. My personal preference is for natural adjective placement, but recognise the need to maintain consistency across docs searches.

I'm willing to try these ideas out and see how ergonomic the API becomes.

**Reply by andrewgryan** [reply]
I've been playing around with the code and spotted a great opportunity for a breaking change. Naturally, this isn't a serious proposal, but I've noticed the number of exports is minimal and I don't want to add extra exports.

```js
// Warning: this breaks everyones code
let { div } = van.tags.html
let { svg, circle } = van.tags.svg
```

I can submit a speculative (WIP) PR if you'd like to peruse the diffs of the library code?

And the custom NS or indeed any type of element route is exposed to the hard-code coder 
```js
// Needs some polish as it's weird to specify the tagName and destructure
let { foo } = van.tags.ns(tagName => document.createElementNS("uri", tagName))
```
That's all my thoughts for now. I can also do the less breaking-change ideas above too if needs be.

**Reply by andrewgryan** [reply]
Further refinement of the above idea, available here: https://github.com/vanjs-org/van/pull/29

**Reply by Tao-VanJS** [reply]
I prefer `const {circle, svg} = van.tagsNS(...)` :-) Breaking changes are not enjoyable to users, thus better to avoid them even before 1.0.

**Reply by andrewgryan** [reply]
Could certainly be a `1.x` candidate as it is a more robust solution, and maps to the idea that tags fit into namespaces.

An alternative approach is to keep `van.tags` with the same behaviour, no breaking change, an carve new ground under a different identifier. `van.elements.svg`, for instance.

**Reply by andrewgryan** [reply]
I've updated #29 to reflect my current thinking. I'd like to take the test suite and type system for a spin to see if I've mucked anything up, but overall I'm happy with the new syntax and could contribute to the docs too if needs be

**Comment by andrewgryan** [comment]
I got curious and looked up how SolidJS implemented their SVG support.

### SolidJS internals
They  use a constant for the namespace string `SVG_NAMESPACE = "http://www.w3.org/2000/svg";` and a flag, `isSVG = false` as an optional argument to their custom `createElement` function. Both of these are **private** implementation details hidden from the user and indeed the curious open source contributor.

https://github.com/solidjs/solid/blob/44a0fdeb585c4f5a3b9bccbf4b7d6c60c7db3ecd/packages/solid/web/src/index.ts#L41

### Exhaustive list of valid SVG elements

To make the difference between HTML and SVG elements invisible to the user they maintain an exhaustive list of SVG/HTML element names, which they import from a third-party package maintained by the SolidJS creator, `babel-plugin-jsx-dom-expressions` which is a whopping *132kb* or 100 times VanJS.

https://github.com/ryansolid/dom-expressions/blob/3bfbff5cbc154aa27f896f73cedc4015ae8877fb/packages/dom-expressions/src/constants.js#L111

I'm not suggesting we follow in SolidJS's footsteps, too much memory overhead. I'd like to point out my latest proposal contains no breaking changes since `vans.tags` is untouched. I'm open to suggestions for better ways to achieve SVG support without burdening the app developer with camelCase or function calls or obscure strings.

**Comment by andrewgryan** [comment]
How about we remove a layer of indirection in the spirit of the current Van API?

``` js
const { div } = van.tags
const { svg, circle } = van.svgs
```

Just a thought, easier to remember I think, and is less code than my previous idea.

**Comment by andrewgryan** [comment]
@Tao-VanJS what do I need to do next to get SVG support over the line?

**Reply by Tao-VanJS** [reply]
Thanks for your contribution! I will get it merged to VanJS's next major release.

**Comment by Tao-VanJS** [comment]
FYI, SVG (and MathML or other HTML elements with `namespaceURI`) is supported in [0.12.0](https://github.com/vanjs-org/van/discussions/53). See https://vanjs.org/tutorial#svg-and-mathml-support.

**Comment by Ashvith10** [comment]
Is there a way to read SVG or raw HTML tags without having to convert it to a function?

**Reply by Tao-VanJS** [reply]
Sorry I don't quite understand the question.

**Reply by Ashvith10** [reply]
@Tao-VanJS sorry, I thought that I was in the convertor repo. After trying that library out, it looks like it can't do what I was looking for - it could probably be achieved with join and eval, but I want to explore simpler options.

Is it possible for me to read HTML/XML(SVG) snippets as inputs, without having to use the online convertor tool? Because doing so isn't portable, and I would have to maintain another line of code. So far, it seems to me that the only possible way to add SVG images is through the `img` tag, but I would also like to explore the option to add inline SVG in the HTML directly.

**Reply by Tao-VanJS** [reply]
Unfortunately, I don't think the converter library supports elements with namespace URIs.

**Reply by Ashvith10** [reply]
@Tao-VanJS Maybe I've not read the documentation thoroughly, but is there a way of allowing raw HTML/XML strings, which can then be converted to innerHTML elements?

**Reply by Ashvith10** [reply]
I've also tried this:

```js
import icon from '../assets/brand.svg?raw'

...

a(
    { href: '/' },
    () => new DOMParser().parseFromString(icon, "image/svg+xml")
)
```

but this does not work:

```console
Uncaught DOMException: Element.append: May not add a Document as a child
    add van.js:90
    tag van.js:115
    Navbar navbar.js:10
    App app.js:6
    <anonymous> main.js:6
van.js:90:4
```

**Reply by Ashvith10** [reply]
I think I have figured this out - all I had to do was read the documentElement attribute from the above code snippet, something like:

```js
import van from '#vanjs'
import icon from '../assets/brand.svg?raw'

const { a, button, div, img, li, nav, ul } = van.tags

export const Navbar = () => {
  const svgElement = new DOMParser().parseFromString(icon, "image/svg+xml").documentElement
 
  return nav(
    a(
      { href: '/' },
      svgElement
    )
  )
}
```

Perhaps, there is a better, alternative way to do this?

**Reply by Ashvith10** [reply]
@Tao-VanJS it looks like what I've done is probably not the right way - because the linter keeps complaining about DOMParser missing.

**Comment by thednp** [comment]
@Ashvith10 You might want to give [this](https://github.com/thednp/vite-plugin-vanjs-svg) a try?

For now
```bash
npm install vite-vanjs-svg
```
